cleverhans/experimental/certification/nn.py

```diff
@@ -74,22 +126,51 @@ def forward_pass(self, vector, layer_index, is_transpose=False, is_abs=False):
     Raises:
       ValueError: if the layer_index is negative or more than num hidden layers
     """
-    if (layer_index < 0 or layer_index > self.num_hidden_layers):
+    if(layer_index < 0 or layer_index > self.num_hidden_layers):
       raise ValueError('Invalid layer index')
 
-    if (self.layer_types[layer_index] == 'ff' or
-        self.layer_types[layer_index] == 'ff_relu'):
-      if is_transpose:
-        return tf.matmul(tf.transpose(self.weights[layer_index]), vector)
-      elif is_abs:
-        return tf.matmul(tf.abs(self.weights[layer_index]), vector)
+    layer_type = self.layer_types[layer_index]
+    # Reshaping the input for convolution appropriately
+    if is_transpose:
+      vector = tf.reshape(vector, self.output_shapes[layer_index])
+      if (layer_type in {'ff', 'ff_relu'}):
+        return_vector = tf.matmul(tf.transpose(self.weights[layer_index]), vector)
+      elif(layer_type in {'conv', 'conv_relu'}):
+        return_vector = tf.nn.conv2d_transpose(vector, self.weights[layer_index],
+                                               output_shape=self.input_shapes[layer_index],
+                                               strides=[1, STRIDE, STRIDE, 1],
+                                               padding=PADDING)
       else:
-        return tf.matmul(self.weights[layer_index], vector)
-
-    raise NotImplementedError('Unsupported layer type: {0}'.format(
-        self.layer_types[layer_index]))
+        raise NotImplementedError('Unsupported layer type: {0}'.format(layer_type))
+      return tf.reshape(return_vector, (self.sizes[layer_index], 1))
+
+    elif is_abs:
+      vector = tf.reshape(vector, self.input_shapes[layer_index])
+      if(layer_type in {'ff', 'ff_relu'}):
+        return_vector = tf.matmul(tf.abs(self.weights[layer_index]), vector)
+      elif(layer_type in {'conv', 'conv_relu'}):
+        return_vector = tf.nn.conv2d(vector,
+                                     tf.abs(self.weights[layer_index]),
+                                     strides=[1, STRIDE, STRIDE, 1],
+                                     padding=PADDING)
+      else:
+        raise NotImplementedError('Unsupported layer type: {0}'.format(layer_type))
+      return tf.reshape(return_vector, (self.sizes[layer_index + 1], 1))
+
+    # Simple forward pass
+    else:
+      vector = tf.reshape(vector, self.input_shapes[layer_index])
+      if(layer_type in {'ff', 'ff_relu'}):
+        return_vector = tf.matmul(self.weights[layer_index], vector)
+      elif (layer_type in {'conv', 'conv_relu'}):
+        return_vector = tf.nn.conv2d(vector, self.weights[layer_index],
+                                     strides=[1, STRIDE, STRIDE, 1],
+                                     padding=PADDING)
+      else:
+        raise NotImplementedError('Unsupported layer type: {0}'.format(layer_type))
+      return tf.reshape(return_vector, (self.sizes[layer_index + 1], 1))
```

Comment:
There are few things in this function which could be improved:
* Support of `is_abs`  and `transpose` at the same time. Right now code won't work properly if both is_abs and transpose are specified
* Reduce code duplication. Currently you call tf.matmul or tf.conv in multiple times.

Few ideas on how to improve the function:
1. you store weight in some local variable and if `is_abs` true then take absolute value of the variable:
    ```
    weight = self.weights[layer_index]
    if is_abs:
      weight = tf.abs(weight)
    ```

2. Based on type of op and whether transpose is needed you choose a function to perform operation:
    ```
    if layer_type in {'ff', 'ff_relu'}:
      return_vector = tf.matmul(weight, vector, transpose_a=is_transpose)
    elif layer_type in {'conv', 'conv_relu'}:
      if is_transpose:
        ...
      else:
        ...
    return tf.reshape(return_vector, ...)
    ```